asteroid/mad.py

```diff
@@ -383,27 +383,80 @@ def _handle_quit(self,_):
       raise SystemExit()
    
    def _handle_set(self,args):
+      (file,lineno) = self.interp_state.lineinfo
       if len(args) == 0:
          # set a breakpoint at the current line
-         (file,lineno) = self.interp_state.lineinfo
          self.line_breakpoints.append((lineno,file))
          return START_DEBUGGER
       elif len(args) == 1:
-         (file,_) = self.interp_state.lineinfo
-         if args[0].isnumeric(): 
-            self.line_breakpoints.append((int(args[0]),file))
+         self._load_program_text(file)
+         if args[0].isnumeric():
+            if self._validate_breakpoint_line(file, int(args[0])):
+               self.line_breakpoints.append((int(args[0]),file))
          else:
-            self.function_breakpoints.append((args[0],file))
+            if self._validate_breakpoint_function(file, args[0]):
+               self.function_breakpoints.append((args[0],file))
          return START_DEBUGGER
       elif len(args) == 2:
-         if args[0].isnumeric(): 
-            self.line_breakpoints.append((int(args[0]),args[1]))
+         self._load_program_text(args[1])
+         if args[0].isnumeric():
+            if self._validate_breakpoint_line(args[1], int(args[0])):
+               self.line_breakpoints.append((int(args[0]),args[1]))
          else:
-            self.function_breakpoints.append((args[0],args[1]))
+            if self._validate_breakpoint_function(args[1], args[0]):
+               self.function_breakpoints.append((args[0],args[1]))
          return START_DEBUGGER
       else:
          print("error: too many arguments to set")
          return START_DEBUGGER
+   
+   def _validate_breakpoint_line(self, fname, lineno):
+      # Load the correct file
+      file_text = self.program_text[fname]
+      # If the line is outside the current file, produce an error message and return
+      if lineno > len(file_text):
+         print("error: cannot place breakpoint on invalid lines")
+         return False
+      # Check if the line is empty
+      line = file_text[lineno-1].strip()
+      if len(line) == 0:
+         print("error: cannot place breakpoints on blank lines")
+         return False
+      # If the line starts with any of these tokens, it is a valid statement and can accept a breakpoint
```

Comment:
This is very fragile, if the language evolves by introducing new tokens this will break and will prevent users from setting break points.  Is there a way to do this in the lexer for example where it is natural for the user to extend collections when introducing new tokens.  See the parser file where we define collection at the beginning.